### PR TITLE
[IMM32] Imm(Lock|Unlock)ImeDpi and ImmSetOpenStatus

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose
Implementing Japanese input...
JIRA issue: [CORE-11700](https://jira.reactos.org/browse/CORE-11700)

## Proposed changes

- Add `IMEDPI` structure to `ntuser.h`.
- Implement `ImmLockImeDpi`/`ImmUnlockImeDpi` functions.
- Rewrite `ImmSetOpenStatus` function.
- Modify some NTUSER function prototypes.
- Modify `imm32.spec`.

### CODE REVIEW

File: dll/win32/imm32/imm.c [hunk 1]
```diff
@@ -2950,40 +3005,134 @@ BOOL WINAPI ImmSetConversionStatus(
...
+                /* lock if possible */
+                if (pImeDpi->dwFlags & IMEDPI_FLAG_UNKNOWN)
+                    pImeDpi = NULL;
+                else
+                    ++(pImeDpi->cLockObj);
+                break;
+            }
+            pImeDpi = pImeDpi->pNext;
+        } while (pImeDpi);
+    }
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:3036]:
```suggestion
    /* Find by hKL */
    for (pImeDpi = g_pImeDpiList; pImeDpi; pImeDpi = pImeDpi->pNext)
    {
        if (pImeDpi->hKL == hKL) /* found */
        {
            /* lock if possible */
            if (pImeDpi->dwFlags & IMEDPI_FLAG_UNKNOWN)
                pImeDpi = NULL;
            else
                ++(pImeDpi->cLockObj);
            break;
        }
    }
```

File: dll/win32/imm32/imm.c [hunk 2]
```diff
@@ -2950,40 +3005,134 @@ BOOL WINAPI ImmSetConversionStatus(
...
+    if (pImeDpi->cLockObj != 0)
+    {
+        RtlLeaveCriticalSection(&g_csImeDpi);
+        return;
+    }
+
+    if ((pImeDpi->dwFlags & IMEDPI_FLAG_UNKNOWN) == 0)
+    {
+        if ((pImeDpi->dwFlags & IMEDPI_FLAG_UNKNOWN2) == 0 ||
+            (pImeDpi->dwUnknown1 & 1) == 0)
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:3067]: Suspicious: how do you know that dwUnknown1 has to have a flag "1", if it's unknown?

File: dll/win32/imm32/imm.c [hunk 3]
```diff
@@ -2950,40 +3005,134 @@ BOOL WINAPI ImmSetConversionStatus(
...
+        {
+            pNext = pEntry->pNext;
+            if (pNext == pImeDpi) /* found */
+            {
+                pEntry->pNext = pImeDpi->pNext;
+                break;
+            }
+            pEntry = pNext;
+        } while (pEntry);
+    }
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:3092]: Suspicious construct, please simplify.
```suggestion
    for (ppEntry = &g_pImeDpiList; *ppEntry; ppEntry = &((*ppEntry)->pNext))
    {
        if (*ppEntry == pImeDpi) /* found */
        {
            *ppEntry = pImeDpi->pNext;
            break;
        }
    }
```

File: dll/win32/imm32/imm.c [hunk 4]
```diff
@@ -2950,40 +3005,134 @@ BOOL WINAPI ImmSetConversionStatus(
...
+    RtlLeaveCriticalSection(&g_csImeDpi);
+    return pImeDpi;
+}
+
+/***********************************************************************
+ *		ImmUnlockImeDpi (IMM32.@)
+ */
+VOID WINAPI ImmUnlockImeDpi(PIMEDPI pImeDpi)
+{
+    PIMEDPI pEntry, pNext;
```
Comment (HBelusca) [inline dll/win32/imm32/imm.c:3047]:
```suggestion
    PIMEDPI* ppEntry;
```
(see below)

**Review by HBelusca — CHANGES_REQUESTED** [review]

File: dll/win32/imm32/imm.c (hunk 1 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: OK.

File: dll/win32/imm32/imm.c (hunk 4 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: OK.

File: dll/win32/imm32/imm.c (hunk 3 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c]: OK.

File: dll/win32/imm32/imm.c (hunk 2 above)
Comment (katahiromz) [reply dll/win32/imm32/imm.c:3062]:
```c
        PIMEDPI pImeDpi = (PIMEDPI)calloc(sizeof(IMEDPI), 1);
        pImeDpi->cLockObj = 1;
        pv = &pImeDpi->dwUnknown1;
        printf("pv: %p\n", pv);
        fflush(stdout);
        DebugBreak();
        (*pImmUnlockImeDpi)(pImeDpi);
```
image
image
My assumption seems wrong. Sorry.

Comment (katahiromz) [reply dll/win32/imm32/imm.c:3062]: UPDATE.
```c
        PIMEDPI pImeDpi = (PIMEDPI)calloc(sizeof(IMEDPI), 1);
        pImeDpi->cLockObj = 1;
        pv = &pImeDpi->dwUnknown1;
        pImeDpi->dwUnknown1 = (argc >= 2 ? atoi(argv[1]) : 0);
        printf("pv: %p\n", pv);
        fflush(stdout);
        DebugBreak();
        (*pImmUnlockImeDpi)(pImeDpi);
```
image
image
We can see the difference.

Comment (katahiromz) [reply dll/win32/imm32/imm.c:3062]: image
image
image

Comment (katahiromz) [reply dll/win32/imm32/imm.c:3062]: `RtlLeaveCriticalSection`, `RtlHeapFree`, `RtlLeaveCriticalSection`, `RtlHeapFree`, ...
These alternate and continue.

Comment (katahiromz) [reply dll/win32/imm32/imm.c:3062]: I have a superb custom debugger that is specialized for x86 reverse engineering. This is also harvest of it.

**Pull request merged**